### PR TITLE
feat: watermark delay in tooltip

### LINKED ISSUE #876: UX to display relative time of watermark to current time
It is very difficult with the timestamp in UTC to figure out how old is the watermark propagation. It will be great if, on mouse over, we can also see the relative time too (e.g., 60sec, 2min, 5h2min, etc.). It is implicit that the timestamp of the watermark will be behind `now()`, so we do not have to make it signed `+` or `-`.

### PR DESCRIPTION
**Pull request by veds-g**
Fixes #876 

- Shows delay of watermark wrt to `time.now()` when hovering on the edge label
- Shows difference in precision from years (~ 5yr 4mo) up to milliseconds (~ 549ms)

image
image
image

### CODE REVIEW

**Review by yhl25 — APPROVED** [review]

File: ui/src/components/pages/Pipeline/partials/Graph/partials/CustomEdge/index.tsx [hunk 1]
```diff
@@ -50,6 +50,51 @@ const CustomEdge: FC<EdgeProps> = ({
...
+    const str = "Delay - ";
+    const fetchWMTime = data?.edgeWatermark?.WMFetchTime;
+    const minWM = getMinWM();
+    let difference =
+      (fetchWMTime ? fetchWMTime : Date.now()) - (minWM == -1 ? 0 : minWM);
+
+    const years = Math.floor(difference / (1000 * 60 * 60 * 24 * 365));
+    difference -= years * (1000 * 60 * 60 * 24 * 365);
+    const months = Math.floor(difference / (1000 * 60 * 60 * 24 * 30));
+    difference -= months * (1000 * 60 * 60 * 24 * 30);
```
Comment (KeranYang) [inline ui/src/components/pages/Pipeline/partials/Graph/partials/CustomEdge/index.tsx:71]: Using 30 days as a month is not always precise. Consider using a third-party library like `moment` to calculate time. Something like below:

```
const startTime = minWM === -1 ? 0 : minWM;
const endTime = fetchWMTime || Date.now();

const duration = moment.duration(moment(endTime).diff(moment(startTime)));
const years = duration.years();
const months = duration.months();
...
```

**Review by KeranYang — CHANGES_REQUESTED** [review]

File: ui/src/components/pages/Pipeline/partials/Graph/partials/CustomEdge/index.tsx (hunk 1 above)
Comment (veds-g) [reply ui/src/components/pages/Pipeline/partials/Graph/partials/CustomEdge/index.tsx:71]: It's not precise but third-party libraries also use a constant for number of days in a year, ...
for example here using moment js, the difference between same dates over `leap-year and non-leap` & `two non-leaps years` is not constant ~ (library considers 366 days / yr)
image

should we use the current implementation or switch to this library instead?

**Comment by vigith** [comment]
Should we use the word `old` rather than `delay`? I am concerned whether it will confuse the users. The problem with `old` and `delay` is that it is relative and there is a reference point, and hence open to interpretation. 

Any suggestions?

**Comment by KeranYang** [comment]
> Should we use the word old rather than delay? I am concerned whether it will confuse the users. The problem with old and delay is that it is relative and there is a reference point, and hence open to interpretation. Any suggestions?

`555ms behind now`?

File: ui/src/components/pages/Pipeline/partials/Graph/partials/CustomEdge/index.tsx [hunk 2]
```diff
@@ -52,47 +53,43 @@ const CustomEdge: FC<EdgeProps> = ({
   }
   const getMinWM = () => {
     if (data?.edgeWatermark?.watermarks) {
+      return Date.now() - 678;
```
Comment (KeranYang) [inline ui/src/components/pages/Pipeline/partials/Graph/partials/CustomEdge/index.tsx:56]: Is this line for test? can we remove it?

Comment (veds-g) [reply ui/src/components/pages/Pipeline/partials/Graph/partials/CustomEdge/index.tsx:56]: yes, missed cleaning that.

**Review by KeranYang — APPROVED** [review]

**Review by mshakira — APPROVED** [review]

**Pull request merged**